magma/smart/smart_bits.py

```diff
@@ -0,0 +1,293 @@
+
+    def __le__(self, other):
+        assert isinstance(other, _SmartExpr)
+        return _SmartComparisonOpExpr(operator.le, self, other)
+
+    def __lshift__(self, other):
+        assert isinstance(other, _SmartExpr)
+        return _SmartShiftOpExpr(operator.lshift, self, other)
+
+    def __invert__(self):
+        return _SmartUnaryOpExpr(operator.invert, self)
+
+
+class _SmartOpExpr(_SmartExpr, metaclass=_SmartOpExprMeta):
+    def __init__(self, determination, op, *args):
+        super().__init__(determination)
+        self._op = op
+        self._args = tuple(args)
+
+    @property
+    def args(self):
+        return self._args
+
+    @property
+    def op(self):
+        return self._op
+
+    def _update(self, *args):
+        self._args = tuple(args)
+
+    def __str__(self):
+        if inspect.isbuiltin(self._op):
+            op_str = self._op.__name__
+        else:
+            op_str = str(self._op)
+        args = ", ".join(str(arg) for arg in self._args)
+        return f"{str(op_str)}({args})"
+
+
+class _SmartExtendOpExpr(_SmartOpExpr):
+
+    class _Extend:
+        def __init__(self, width):
+            self._width = width
+
+        def __call__(self, operand):
+            return operand.zext(self._width)
+
+        def __str__(self):
+            return f"Extend[{self._width}]"
+
+    def __init__(self, width, operand):
+        extend = _SmartExtendOpExpr._Extend(width)
+        super().__init__(Determination.SELF_DETERMINED, extend, operand)
+
+
+def _extend_if_needed(expr, to_width):
+    diff = expr._width_ - to_width
+    assert diff <= 0
+    if diff == 0:
+        return expr
+    expr = _SmartExtendOpExpr(-diff, expr)
+    expr._width_ = to_width
+    return expr
+
+
+class _SmartBinaryOpExpr(_SmartOpExpr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(Determination.CONTEXT_DETERMINED,
+                         op, loperand, roperand)
+
+    def resolve(self, context):
+        for arg in self._args:
+            arg.resolve(context)
+        to_width = context.max_width()
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = to_width
+
+
+class _SmartUnaryOpExpr(_SmartOpExpr):
+    def __init__(self, op, operand):
+        super().__init__(Determination.CONTEXT_DETERMINED, op, operand)
+
+    def resolve(self, context):
+        raise NotImplementedError()
+
+
+class _SmartComparisonOpExpr(_SmartOpExpr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(Determination.SELF_DETERMINED,
+                         op, loperand, roperand)
+
+    def resolve(self, context):
+        context = Context(None, self)
+        for arg in self._args:
+            arg.resolve(context)
+        to_width = max(arg._width_ for arg in self._args)
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = 1
```

Comment:
Seems like there might be some way to share some of the common code with _SmartBinaryOpExpr.resolve

Reply:
I actually racked my brain on this for a while. Couldn't find a good abstraction, though the structure of the two are very similar. Probably implies that the overall pattern smells a bit (i.e. should be revisited). Open to suggestions if a refactoring is evident to you.